bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java

```diff
@@ -39,7 +50,78 @@ public BigQueryClientFactory(
     this.bqConfig = bqConfig;
   }
 
-  public BigQueryReadClient createBigQueryReadClient(Optional<String> endpoint) {
+  public BigQueryReadClient getBigQueryReadClient() {
+    synchronized (readClientMap) {
+      if (!readClientMap.containsKey(this)) {
+        BigQueryReadClient bigQueryReadClient =
+            createBigQueryReadClient(this.bqConfig.getEndpoint());
+        Runtime.getRuntime()
+            .addShutdownHook(new Thread(() -> shutdownBigQueryReadClient(bigQueryReadClient)));
+        readClientMap.put(this, bigQueryReadClient);
+      }
+    }
+
+    return readClientMap.get(this);
+  }
+
+  public BigQueryWriteClient getBigQueryWriteClient() {
+    synchronized (writeClientMap) {
+      if (!writeClientMap.containsKey(this)) {
+        BigQueryWriteClient bigQueryWriteClient = createBigQueryWriteClient();
+        Runtime.getRuntime()
+            .addShutdownHook(new Thread(() -> shutdownBigQueryWriteClient(bigQueryWriteClient)));
+        writeClientMap.put(this, bigQueryWriteClient);
+      }
+    }
+
+    return writeClientMap.get(this);
+  }
+
+  @Override
+  public int hashCode() {
+    // Here, credentials is an instance of GoogleCredentials which can be one out of
+    // GoogleCredentials, UserCredentials, ServiceAccountCredentials, ExternalAccountCredentials or
+    // ImpersonatedCredentials (See the class BigQueryCredentialsSupplier which supplies these
+    // Credentials). Subclasses of the abstract class ExternalAccountCredentials do not have the
+    // hashCode method defined on them and hence we only compute the hashcode using
+    // userAgentHeaderProvider and bqConfig here. In the equals implementation, we serialize and
+    // compare byte arrays if either of the credentials are instances of ExternalAccountCredentials
+    if (credentials instanceof ExternalAccountCredentials) {
```

Comment:
`hashCode()` input should be the same as `equals()`, so please create a hashCode for `ExternalAccountCredentials` as well. Th e best is to move all credentials handling code (including the equals related methods below) to an independent utility class - none of this functionality is unique to `BigQueryClientFactory`